wallet/src/states.rs

```diff
@@ -64,319 +53,160 @@ impl<K, S> State<K, S> {
-        let mut state = Box::new(State::new(key, state, Status::Confirmed));
-        let key_ref = KeyRef(&state.key);
-        let head: *mut State<K, S> = &mut *state;
-        let tail: *mut State<K, S> = &mut *state;
+        let state = StateRef::new(key.clone(), state, Status::Confirmed);
+        let mut states = LinkedHashMap::new();
+        states.insert(key, state);
 
-        let mut map = HashMap::with_capacity(12);
-        map.insert(key_ref, state);
-
-        Self { map, head, tail }
+        Self { states }
     }
 
     /// check wether the given state associate to this key is present
     /// in the States
     pub fn contains<Q: ?Sized>(&self, key: &Q) -> bool
     where
-        KeyRef<K>: Borrow<Q>,
-        Q: Hash + Eq,
-    {
-        self.map.contains_key(key)
-    }
-
-    /// get the underlying State associated to the given key
-    pub fn get<Q: ?Sized>(&self, key: &Q) -> Option<(&S, Status)>
-    where
-        KeyRef<K>: Borrow<Q>,
+        K: Borrow<Q>,
         Q: Hash + Eq,
     {
-        self.map.get(key).map(|s| (&s.state, s.status))
+        self.states.contains_key(key)
     }
 
     /// push a new **unconfirmed** state in the States
     pub fn push(&mut self, key: K, state: S) {
-        let mut state = Box::new(State::new(key, state, Status::Pending));
-        let key_ref = KeyRef(&state.key);
+        let state = StateRef::new(key.clone(), state, Status::Pending);
 
-        state.prev = self.tail;
-        unsafe { (*self.tail).next = state.as_mut() };
-        self.tail = state.as_mut();
-
-        assert!(self.map.insert(key_ref, state).is_none());
+        assert!(self.states.insert(key, state).is_none());
     }
 
     pub fn confirm<Q: ?Sized>(&mut self, key: &Q)
     where
-        KeyRef<K>: Borrow<Q>,
+        K: Borrow<Q>,
         Q: Hash + Eq,
     {
-        if let Some(state) = self.map.get_mut(key) {
-            let state = &mut (*state) as &mut State<K, S>;
+        if let Some(state) = self.states.get_mut(key) {
             state.confirm();
         }
 
-        while self.pop_legacy_confirmed() {}
+        self.pop_old_confirmed_states()
     }
 
-    fn pop_legacy_confirmed(&mut self) -> bool {
-        let current = unsafe { &mut (*self.head) as &mut State<K, S> };
-        debug_assert!(current.confirmed());
-
-        if let Some(next) = unsafe { current.next.as_mut() } {
-            if next.confirmed() {
-                let current = self.map.remove(&current.key);
+    fn pop_old_confirmed_states(&mut self) {
+        loop {
+            let (key, state) = self.states.pop_front().unwrap();
 
-                self.head = current.expect("head reference is not in map").next;
-                next.prev = std::ptr::null_mut();
+            let finished = self
+                .states
+                .front()
+                .map(|(_, state)| state.is_pending())
+                .unwrap_or(true);
 
-                return true;
+            if finished {
```

Comment:
What about `self.states.iter().skip(1).map(|(_, state)| state.is_pending()).unwrap_or(true)` and then popping only if that value is false?

Reply:
I like that, but I think that would be the same as first doing an iteration round to find the stopping point, and then a round of popping, right? In the number of operations I mean.

Something like (untested):
```rust
let confirmed_count = self.states.iter().take_while(|(_, state)| state.is_confirmed()).map(|_| 1).sum();

while confirmed_count > 1 {
    self.states.pop_front()
}
```

I was initially doing this without popping (although in a worst way)... And then changed it, but maybe avoiding reinserting is better, as we can remove the `Clone` bound too.

Reply:
Yep, and I think it's more readable your way (although possibly less cache friendly, but this is not important at all here)

Reply:
I ended up writing it your way (although negated), and I ended up actually liking it more